### PR TITLE
8315890: Attempts to load from nullptr in instanceKlass.cpp and unsafe.cpp

### PR DESCRIPTION
**Pull request by matias9927**
Calls in instanceKlass.cpp and unsafe.cpp try to call an atomic load on method calls that could return nullptr. This patch ensures that nullptr is not passed into the load. 

In `print_as_native_pointer` in archiveBuilder, `source_obj_to_requested_obj` should not be able to return `nullptr` as the result is immediately cast to an oop which cascades down to the failure reported in  `get_volatile()` in `unsafe.cpp`. Placing an assert close to the top of this call stack should prevent this from happening and will better indicate the source of an unexpected `nullptr` should it occur.

Verified with tier1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315890](https://bugs.openjdk.org/browse/JDK-8315890): Attempts to load from nullptr in instanceKlass.cpp and unsafe.cpp (**Bug** - P3)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16405/head:pull/16405` \
`$ git checkout pull/16405`

Update a local copy of the PR: \
`$ git checkout pull/16405` \
`$ git pull https://git.openjdk.org/jdk.git pull/16405/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16405`

View PR using the GUI difftool: \
`$ git pr show -t 16405`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16405.diff">https://git.openjdk.org/jdk/pull/16405.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16405#issuecomment-1783213872)

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Looks good.

**Review by calvinccheung — APPROVED** [review]
LGTM

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 1]
```diff
@@ -2502,10 +2502,10 @@ void InstanceKlass::clean_implementors_list() {
     assert (ClassUnloading, "only called for ClassUnloading");
     for (;;) {
       // Use load_acquire due to competing with inserts
-      InstanceKlass* impl = Atomic::load_acquire(adr_implementor());
+      InstanceKlass* volatile* iklass = adr_implementor();
+      InstanceKlass* impl = (iklass != nullptr) ? Atomic::load_acquire(iklass) : nullptr;
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:2506]: This looks very klunky as we do a raw read, check it for null then re-read with acquire semantics. Cleaner IMO to do a raw read followed by a raw `OrderAccess::acquire()` and no need for a null check.

File: src/hotspot/share/prims/unsafe.cpp [hunk 2]
```diff
@@ -231,6 +231,7 @@ class MemoryAccess : StackObj {
 
   T get_volatile() {
     GuardUnsafeAccess guard(_thread);
+    assert(addr() != nullptr, "Attempting to load from null pointer");
```
Comment (dholmes-ora) [inline src/hotspot/share/prims/unsafe.cpp:234]: I don't see how `addr()` can be null unless `_obj` was null - which would be a usage error. So asserting `_obj != nullptr` in the constructor would seem better to me. I mean no point checking `addr()` here but not in other functions where we dereference it!

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Review by coleenp — APPROVED** [review]
Yes, this looks better.  This was the source of the nullptr.

**Review by calvinccheung — APPROVED** [review]
Updated changes look good.

**Review by dholmes-ora — COMMENTED** [review]
Not sure why it needed to be lifted out of Unsafe. The issue description should be updated now.

**Comment by matias9927** [comment]
> Not sure why it needed to be lifted out of Unsafe. The issue description should be updated now.

Sorry, I meant to explain this change in response to your comments. I'll explain here:

The call stack that results in the issue shown in `get_volatile()` starts near where the new assert is placed. When discussing with @coleenp, we decided that placing the assert at the source of the nullptr would be a better indicate the problem should it arise in the code. The description has been updated with more detail.

**Review by dholmes-ora — APPROVED** [review]
Okay. Thanks

**Comment by matias9927** [comment]
Thanks for the reviews @coleenp , @calvinccheung , and @dholmes-ora!
/integrate

**Pull request closed without merging**